checker/src/main/java/org/checkerframework/checker/confidential/ConfidentialTransfer.java

```diff
@@ -0,0 +1,128 @@
+import org.checkerframework.dataflow.cfg.node.StringConcatenateNode;
+import org.checkerframework.framework.flow.CFAbstractAnalysis;
+import org.checkerframework.framework.flow.CFStore;
+import org.checkerframework.framework.flow.CFTransfer;
+import org.checkerframework.framework.flow.CFValue;
+import org.checkerframework.framework.type.QualifierHierarchy;
+import org.checkerframework.javacutil.AnnotationUtils;
+
+/** The transfer class for the Confidential Checker. */
+public class ConfidentialTransfer extends CFTransfer {
+
+  /** The Confidential type factory. */
+  protected final ConfidentialAnnotatedTypeFactory atypeFactory;
+
+  /** The Confidential qualifier hierarchy. */
+  protected final QualifierHierarchy qualHierarchy;
+
+  /**
+   * Create a new ConfidentialTransfer.
+   *
+   * @param analysis the corresponding analysis
+   */
+  public ConfidentialTransfer(CFAbstractAnalysis<CFValue, CFStore, CFTransfer> analysis) {
+    super(analysis);
+    atypeFactory = (ConfidentialAnnotatedTypeFactory) analysis.getTypeFactory();
+    qualHierarchy = atypeFactory.getQualifierHierarchy();
+  }
+
+  /**
+   * Enforces Confidential String concatenation rules:
+   *
+   * <ul>
+   *   <li>(Confidential + NonConfidential) returns Confidential (commutatively);
+   *   <li>(Confidential + Confidential) returns Confidential;
+   *   <li>(NonConfidential + NonConfidential) returns NonConfidential;
+   *   <li>UnknownConfidential dominates other types in concatenation;
+   *   <li>Non-bottom types dominate BottomConfidential in concatenation.
+   * </ul>
+   */
+  @Override
+  public TransferResult<CFValue, CFStore> visitStringConcatenate(
+      StringConcatenateNode n, TransferInput<CFValue, CFStore> p) {
+    TransferResult<CFValue, CFStore> result = super.visitStringConcatenate(n, p);
+    return stringConcatenation(n.getLeftOperand(), n.getRightOperand(), p, result);
+  }
+
+  /**
+   * Determines the type of a string concatenation.
+   *
+   * @param leftOperand the left operand to be concatenated
+   * @param rightOperand the right operand to be concatenated
+   * @param p the input abstract values
+   * @param result the result abstract values
+   * @return the resulting type of the string concatenation operation.
+   */
+  public TransferResult<CFValue, CFStore> stringConcatenation(
+      Node leftOperand,
+      Node rightOperand,
+      TransferInput<CFValue, CFStore> p,
+      TransferResult<CFValue, CFStore> result) {
+    AnnotationMirror resultAnno =
+        createAnnotationForStringConcatenation(leftOperand, rightOperand, p);
+    return recreateTransferResult(resultAnno, result);
+  }
+
+  /**
+   * Creates an annotation for a result of string concatenation.
+   *
+   * @param leftOperand the left operand to be concatenated
+   * @param rightOperand the right operand to be concatenated
+   * @param p the input abstract values
+   * @return the resulting AnnotationMirror of the string concatenation operation
```

Comment:
This method may also return `null`.  Please document the conditions under which that happens.